### PR TITLE
add support for custom formats (mantis 6452)

### PR DESCRIPTION
**Pull request by ghost**
This is an implementation of the feature suggested in [mantis 6452](http://caml.inria.fr/mantis/view.php?id=6452).

The idea is to allow ppx rewriters to extend the syntax of format strings in a simple way. The patch allows to embed arbitrary functions into a format by adding a `Custom` constructor to the type `CamlinternalFormatBasics.fmt`:

``` ocaml
type ('a, 'b, 'c, 'd) custom_arity =
  | Custom_zero : ('a, 'b, 'a, 'b) custom_arity
  | Custom_succ : ('a, 'b, 'c, 'd) custom_arity ->
    ('a, 'b, 'x -> 'c, 'x -> 'd) custom_arity

type ('a, 'b, 'c, 'd, 'e, 'f) fmt =
...
| Custom :
    ('c, 'a, 'x, 'y) custom_arity * ('b -> 'x) * ('a, 'b, 'c, 'd, 'e, 'f) fmt ->
    ('y, 'b, 'c, 'd, 'e, 'f) fmt
```

One example of project already extending the syntax of format string is [custom_printf](https://github.com/janestreet/custom_printf). It allows to write expression like this:

``` ocaml
printf !"The time is %{sexp:Time.t}." time
```

behaving the same as:

``` ocaml
printf "The time is %s." (Sexp.to_string_hum (Time.sexp_of_t time)))
```

At the moment the only way to do such a rewriting is to modify the call site where the format string appears. This make the extension quite complicated and not very robust; it is not possible have an extended format string by itself for instance. With the proposed patch the ppx rewriter can simply _map_ the format string. In this example it would rewrite it to:

``` ocaml
String_literal ("The time is ",
  Custom (Custom_succ Custom_zero,
          (fun () time -> Sexp.to_string_hum (Time.sexp_of_t time)),
          String_literal (".", End_of_format))
```

### CODE REVIEW

File: stdlib/camlinternalFormat.ml [hunk 1]
```diff
@@ -537,6 +544,16 @@ let bprint_fmt buf fmt =
     | Theta rest ->
       buffer_add_char buf '%'; bprint_ignored_flag buf ign_flag;
       buffer_add_char buf 't'; fmtiter rest false;
+    | Custom (arity, _, rest) ->
+      let n = int_of_custom_arity arity in
+      if n > 0 then begin
+        buffer_add_char buf '%'; bprint_ignored_flag buf ign_flag;
+        buffer_add_char buf '?';
+        for _i = 1 to n - 1 do
+          buffer_add_string buf "%?"
```
Comment (gasche) [inline stdlib/camlinternalFormat.ml]: This is the only part of the patch that might be wrong. Shouldn't we use `bprint_ignored_flag` on all `n` parameters?

**Comment by gasche** [comment]
I like the patch. @bvaugon, any opinion?

File: stdlib/camlinternalFormat.ml (hunk 1 above)
Comment (diml) [inline stdlib/camlinternalFormat.ml]: I wasn't sure about this part, I think `ign_flag=true` is actually impossible here. But for uniformity with the other cases it make sense to use `bprint_ignored_flag` everywhere indeed. I'll update the patch.

Comment (gasche) [inline stdlib/camlinternalFormat.ml]: Yes, the flag would only appear if you (1) added an "ignored" counterpart of Custom_arity and (2) generated code using it. I think it's reasonable to assume that you may want to do that in the future, eg. to support `%_{sexp:Time.t}`.

**Comment by diml** [comment]
Any news on this? I'd like to apply the patch if nobody objects.

**Comment by planar** [comment]
Shouldn't it be documented somewhere? I'm not sure where, but I think such a useful feature should have documentation. Maybe in a document about writing ppx processors. Or at least in code comments...

Also, when you're printing the custom format, you print it as `%?`, but when you mention it in an error message, you display it as `<custom>`. I see some potential for confusion here.

Finally, would it be interesting (and possible) to give it a concrete syntax?

**Comment by Drup** [comment]
For some reason, I missed this PR, this is really great ! Thanks @diml.

**Comment by whitequark** [comment]
Extremely nice.

**Comment by diml** [comment]
> Shouldn't it be documented somewhere? I'm not sure where, but I think such a useful feature should have documentation. Maybe in a document about writing ppx processors. Or at least in code comments...

Sure. I am not sure what is the best place to put this. I can write a blog post with a complete example.

> Also, when you're printing the custom format, you print it as %?, but when you mention it in an error message, you display it as <custom>. I see some potential for confusion here.

OK, I changed the error message in scanf.ml to `"scanf: bad conversion \"%?\" (custom converter)"`.

> Finally, would it be interesting (and possible) to give it a concrete syntax?

I guess we could but most suitable syntaxes are already used. In custom-printf we just forbid the rarely used `%{ ... %}`  syntax.

**Comment by dbuenzli** [comment]
Le jeudi, 5 mars 2015 à 11:55, Jérémie Dimino a écrit :
> Sure. I am not sure what is the best place to put this. I can write a blog post with a complete example.

In the ocamldoc documentation about formatting strings. Blog posts get wiped from the internet, but you will still have the mli you need to use.

Best,

Daniel

**Comment by diml** [comment]
I wrote a small comment about custom converters in the description of format strings.

**Comment by gasche** [comment]
Merged in trunk. I moved the documentation of custom formats to CamlinternalFormatBasics. I don't think it belongs to end-user-visible documentation which is more about the particular format syntax that the Printf module supports. And let me take this as an opportunity to insist again that the format datatype is internal and will change in future versions.

**Pull request closed without merging**